warp/Network/Wai/Handler/Warp.hs

```diff
@@ -371,3 +391,10 @@ setHTTP2Disabled y = y { settingsHTTP2Enabled = False }
 -- Since 3.0.10
 pauseTimeout :: Request -> IO ()
 pauseTimeout = fromMaybe (return ()) . Vault.lookup pauseTimeoutKey . vault
+
+-- | Getting file information from the inside cache.
+--   If file information is not available, an IO error is thrown.
+--
+-- Since 3.1.10
+getFileInfo :: Request -> FilePath -> IO FileInfo
+getFileInfo = fromMaybe (\_ -> throwIO (userError "getFileInfo")) . Vault.lookup getFileInfoKey . vault
```

Comment:
For (2), I think that you misunderstand.

`getFileInfo` tries to obtain the file information of the target file. This is main behavior. 

Using the cache or registering to the cache is just for performance purpose only.

In other words, this is just a high-performance `stat(2)` syscall. So, it's strange to have `Maybe` for its semantics.